### Chapter5_Queation 9

We will now consider the <font color="brown">Boston</font> housing data set, from the <font color="brown">ISLP</font> library.

In [1]:
from ISLP import load_data
Boston=load_data("Boston")
Boston.head()

crim    zn  indus  chas    nox     rm   age     dis  rad  tax  ptratio  \
0  0.00632  18.0   2.31     0  0.538  6.575  65.2  4.0900    1  296     15.3   
1  0.02731   0.0   7.07     0  0.469  6.421  78.9  4.9671    2  242     17.8   
2  0.02729   0.0   7.07     0  0.469  7.185  61.1  4.9671    2  242     17.8   
3  0.03237   0.0   2.18     0  0.458  6.998  45.8  6.0622    3  222     18.7   
4  0.06905   0.0   2.18     0  0.458  7.147  54.2  6.0622    3  222     18.7   

   lstat  medv  
0   4.98  24.0  
1   9.14  21.6  
2   4.03  34.7  
3   2.94  33.4  
4   5.33  36.2

(a) Based on this data set, provide an estimate for the population mean of <font color="brown">medv</font>. Call this estimate $\hat{µ}$.

In [6]:
print(f"estimate of population mean of medv: {Boston["medv"].mean():.4f}")

estimate of population mean of medv: 22.5328


(b) Provide an estimate of the standard error of $\hat{µ}$. Interpret this result.<br>
_Hint: We can compute the standard error of the sample mean by dividing the sample standard deviation by the square root of the number of observations._

In [5]:
import numpy as np
y = Boston["medv"].values

# Compute standard deviation
std = np.std(y, ddof=1)

# Number of observations
n = len(y)

# Standard error of sample mean
se_mean = std / np.sqrt(n)

print(f"Sample standard deviation: {std:.4f}")
print(f"Standard error of mean: {se_mean:.4f}")

Sample standard deviation: 9.1971
Standard error of mean: 0.4089


(c) Now estimate the standard error of $\hat{µ}$ using the bootstrap. How does this compare to your answer from (b)?

In [10]:
s = np.std(y, ddof=1)
se_analytic = s / np.sqrt(n)

# Bootstrap setting
B = 10000
rng = np.random.default_rng(314657018)  
boot_means = np.empty(B)

for b in range(B):
    sample = rng.choice(y, size=n, replace=True)
    boot_means[b] = sample.mean()

se_boot = np.std(boot_means, ddof=1)

print(f"n = {n}")
print(f"Sample SD (s): {s:.4f}")
print(f"Analytic SE = s / sqrt(n): {se_analytic:.4f}")
print(f"Bootstrap SE (B={B}): {se_boot:.4f}")
print(f"Absolute difference: {abs(se_boot - se_analytic):.6f}")
print(f"Relative difference: {abs(se_boot - se_analytic)/se_analytic:.4%}")

n = 506
Sample SD (s): 9.1971
Analytic SE = s / sqrt(n): 0.4089
Bootstrap SE (B=10000): 0.4067
Absolute difference: 0.002144
Relative difference: 0.5245%


(d) Based on your bootstrap estimate from (c), provide a 95 % confidence interval for the mean of <font color="brown">medv</font>. Compare it to the results obtained by using <font color="brown">Boston['medv'].std()</font> and the two standard error rule (<font color="blue">3.9</font>).

_Hint: You can approximate a 95% confidence interval using the formula $[\hat{µ}− 2SE(\hat{µ}),\hat{µ} + 2SE(\hat{µ})]$._

In [15]:
mu_hat = y.mean()

# 95% CI approximate using +/- 2 * SE_boot
ci_lower = mu_hat - 2 * se_boot
ci_upper = mu_hat + 2 * se_boot

print(f"95% CI by bootstrap (mu_hat ± 2 * SE_boot): [{ci_lower:.4f}, {ci_upper:.4f}]")


95% CI by bootstrap (mu_hat ± 2 * SE_boot): [21.7194, 23.3462]


(e) Based on this data set, provide an estimate, $\hat{µ}_{med}$, for the median value of <font color="brown">medv</font> in the population.

In [17]:
print(f"estimate of population median of medv: {Boston["medv"].median():.4f}")

estimate of population median of medv: 21.2000


(f) We now would like to estimate the standard error of $\hat{µ}_{med}$. Unfortunately, there is no simple formula for computing the standard
error of the median. Instead, estimate the standard error of the median using the bootstrap. Comment on your findings.

In [18]:
# bootstrap
B = 10000
rng = np.random.default_rng(314657018)
boot_medians = np.empty(B)
for b in range(B):
    sample = rng.choice(y, size=n, replace=True)
    boot_medians[b] = np.median(sample)

se_median_boot = np.std(boot_medians, ddof=1)

# optional: bootstrap percentile 95% CI for the median
lower_pct, upper_pct = np.percentile(boot_medians, [2.5, 97.5])

print(f"n = {n}")
print(f"Bootstrap SE of median (B={B}): {se_median_boot:.4f}")
print(f"Bootstrap 95% percentile CI for median: [{lower_pct:.4f}, {upper_pct:.4f}]")

n = 506
Bootstrap SE of median (B=10000): 0.3751
Bootstrap 95% percentile CI for median: [20.5000, 21.8500]


<font color="blue">**Answer:**
I estimate the standard error of (\hat\mu_{med}) (the sample median of `medv`) by bootstrap (resampling the 506 observations with replacement (B=10,000) times), and computing the standard deviation of the bootstrap medians. The code above returns the bootstrap standard error `se_median_boot` and a bootstrap percentile 95% interval for the median.</font>

(g) Based on this data set, provide an estimate for the tenth percentile of <font color="brown">medv</font> in Boston census tracts. Call this quantity $\hat{µ}_{0.1}$. <br>(You can use the <font color="brown">np.percentile()</font> function.) 

In [19]:
mu_0_1 = np.percentile(y, 10)
print(f"Estimated 10th percentile (mu_0.1): {mu_0_1:.4f}")

Estimated 10th percentile (mu_0.1): 12.7500


(h) Use the bootstrap to estimate the standard error of $\hat{µ}_{0.1}$. Comment on your findings.

In [20]:
B = 10000
rng = np.random.default_rng(314657018)

boot_p10 = np.empty(B)
n = len(y)

for b in range(B):
    sample = rng.choice(y, size=n, replace=True)
    boot_p10[b] = np.percentile(sample, 10)

se_p10 = np.std(boot_p10, ddof=1)

print(f"Bootstrap SE of 10th percentile estimate: {se_p10:.4f}")

Bootstrap SE of 10th percentile estimate: 0.5046
